test/tstQueryTreeCallbackKernelQuery.cpp

```diff
@@ -0,0 +1,71 @@
+/****************************************************************************
+ * Copyright (c) 2023 by the ArborX authors                                 *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include "ArborX_EnableDeviceTypes.hpp" // ARBORX_DEVICE_TYPES
+#include <ArborX.hpp>
+
+#include <boost/test/unit_test.hpp>
+
+#include <numeric>
+#include <vector>
+
+BOOST_AUTO_TEST_SUITE(KerneliQueryCallbacks)
+
+struct IntersectionCallback
+{
+  int query_index;
+  bool &success;
```

Comment:
This is a gray area.

Reply:
Does this violate any of our semantics? Why do you call it "gray area"?

Reply:
For me, being able to capture variables by reference is what makes this interface quite useful/powerful.